core/Symbols.cc

```diff
@@ -818,6 +818,70 @@ SymbolRef Symbol::topAttachedClass(const GlobalState &gs) const {
     return classSymbol;
 }
 
+void Symbol::recordSealedSubclass(MutableContext ctx, SymbolRef subclass) {
+    ENFORCE(this->isClassSealed(), "Class is not marked sealed: {}", this->show(ctx));
+    ENFORCE(subclass.exists(), "Can't record sealed subclass for {} when subclass doesn't exist", this->show(ctx));
+    ENFORCE(subclass.data(ctx)->isClass(), "Sealed subclass {} must be class", subclass.show(ctx));
+
+    auto classOfSubclass = subclass.data(ctx)->singletonClass(ctx);
+    auto sealedClassesList =
+        this->lookupSingletonClass(ctx).data(ctx)->findMember(ctx, core::Names::sealedClassesList());
+
+    auto data = sealedClassesList.data(ctx);
+    ENFORCE(data->resultType != nullptr, "Should have been populated in namer");
+    auto appliedType = cast_type<AppliedType>(data->resultType.get());
+    ENFORCE(appliedType != nullptr, "sealedClassesList should always be AppliedType");
+    ENFORCE(appliedType->klass == core::Symbols::Array(), "sealedClassesList should always be Array");
+    auto currentClasses = appliedType->targs[0];
+    // Abusing T.any to be list cons, with T.noreturn as the empty list.
+    // (Except it's not, because Types::lub is too smart, and drops the T.noreturn to prevent allocating a T.any)
+    appliedType->targs[0] = Types::any(ctx, currentClasses, make_type<ClassType>(classOfSubclass));
+}
+
+const InlinedVector<Loc, 2> &Symbol::sealedLocs(const GlobalState &gs) const {
+    ENFORCE(this->isClassSealed(), "Class is not marked sealed: {}", this->show(gs));
+    auto sealedClassesList = this->lookupSingletonClass(gs).data(gs)->findMember(gs, core::Names::sealedClassesList());
+    auto &result = sealedClassesList.data(gs)->locs();
+    ENFORCE(result.size() > 0);
+    return result;
+}
+
+TypePtr Symbol::sealedSubclasses(const Context ctx) const {
+    ENFORCE(this->isClassSealed(), "Class is not marked sealed: {}", this->show(ctx));
```

Comment:
can we add some enforce that effectively verifies that computation of all potential sealed classes has terminated?

Reply:
maybe check that global state is frozen?
or if linearization has completed? both of these can be good proxies